text/0000-variadic-generics.md

```diff
@@ -0,0 +1,444 @@
+
+Variadic generic parameters always take the form `...$ty`.
+
+### Defining types
+[defining-types]: #defining-types
+
+A `...T` in generic formal type parameters, where `...T` is the trailing
+type will define the type as variadic.
+
+```rust
+struct S<...T>(...T);
+
+S();
+S(42);
+S(12, 13, 17)
+```
+
+### Defining functions
+[defining-functions]: #defining-functions
+
+Similar to defining types the same syntax may be used in a functions generic
+parameters to define a variable length list of types.
+
+```rust
+fn bar<...T>(/* arguments */);
+```
+
+### Capturing multiple arguments
+[capturing-multiple-arguments]: #capturing-multiple-arguments
+
+In the functions arguments the pattern `...args: T` (where `T` is the name of
+the variadic list of types and `args` is the argument name) will be allowed
+to define a function which takes a variable number of arguments at the position
+of `args`.
+
+```rust
+fn bar<...T>(...args: T) -> R {...}
+
+bar();
+bar(42);
+bar(12, 13, 17);
+
+fn baz<...T>(n: i32, ...args: T) -> R { ... }
+
+baz(0);
+baz(1, 42);
+```
+
+### Unpacking variadic arguments
+[unpacking-variadic-arguments]: unpacking-variadic-arguments
+
+In order to unpack the variable length list of arguments in the function
+body, the pattern `...args` may be used in the function body.
+
+```rust
+fn f<...T>(...args: T) -> R {
+    // Unpack args and pass them to the function g
+    g(...args)
+}
+```
+
+### Defining bounds for `T`
+[defining-bounds-for-t]: #defining-bounds-for-t
+
+A type bound for each type in the variadic list of types `...T` may be
+defines with the syntax `...T: Trait`. This indicates that each type type
+in `T` must satisfy the bound `Trait`.
+
+```rust
+fn example<...T: Clone>(...args: T);
+```
+
+Now that the types in `T` can be bound and the list of arguments may be unpacked
+it becomes possible to iterate over the list of arguments and call trait methods
+on each argument in the list.
+
+```rust
+fn example<...T: Clone>(...args: T) {
+    // All types in T implement clone, so we can use
+    // the clone method on the unpacked args
+    f(...args.clone());
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+When working with variadics it is helpful to be able to:
+
+ - unpack the arguments
+ - recurse over the arguments
+ - fold over the arguments
+
+## Recursion
+[recursion]: #recursion
+
+In order to allow recursing over the list there needs to be a way to
+extract an argument (the first argument in most cases). The addition
+of the pattern `...arg` to represent a argument that "captures"
+multiple arguments can be employed to allow this. When used as an
+l-value pattern to an expression, the variable `...arg` in a tuple
+pattern will consume the remainder of elements in the variadic list
+of arguments. For example, a simple recursive function might look
+something like the following:
+
+```rust
+fn recurse<...T>(...args: T) {
+   let (head, ...tail) = args;
```

Comment:
We already have `let (head, ..) = args;`, so this should just be `let (head, tail @ ..) = args;`.